Filrb/Internal/Invariants/RBInv.lean

```diff
@@ -0,0 +1,162 @@
+import Filrb.Internal.Raw
+import Filrb.Internal.Invariants.BST
+
+/-!
+This module proves that our functionality preserves the `RedBlackTree` invariant,
+which is a combination of the `ChildInv` and the `HeightInv` invariant.
+-/
+
+namespace Filrb
+namespace Internal
+
+namespace Raw
+
+variable [Preorder α] [Ord α] [LawfulOrd α]
+
+/--
+The child invariant for red black trees: Red nodes must have black children.
+-/
+inductive ChildInv : Raw α → Prop where
+  | nil : ChildInv .nil
+  | black (hleft : ChildInv left) (hright : ChildInv right) : ChildInv (.node left data .black right)
+  | red (hleft1 : ChildInv left) (hleft2 : left.rootColor = .black) (hright1 : ChildInv right)
+        (hright2 : right.rootColor = .black) : ChildInv (.node left data .red right)
+
+attribute [pp_nodot] ChildInv
+
+/--
+A weaker child invariant for red black trees,
+where only the childs of a node have to preserve the invariant.
+-/
+def ChildInv2 (t : Raw α) : Prop :=
+  ChildInv (paintColor .black t)
+
+attribute [pp_nodot] ChildInv2
+
+omit [Preorder α] [Ord α] [LawfulOrd α] in
+@[simp]
+theorem childInv_nil : ChildInv (.nil : Raw α) := ChildInv.nil
+
+omit [Preorder α] [Ord α] [LawfulOrd α] in
+@[simp]
+theorem childInv2_nil : ChildInv2 (.nil : Raw α) := ChildInv.nil
+
+omit [Preorder α] [Ord α] [LawfulOrd α] in
+@[simp]
+theorem childInv_red_node {l r : Raw α} :
+    ChildInv (.node l d .red r) ↔ ChildInv l ∧ (l.rootColor = .black) ∧ ChildInv r ∧ (r.rootColor = .black) := by
+  constructor
+  · intro h
+    cases h
+    simp_all
+  · rintro ⟨_, _, _, _⟩
+    apply ChildInv.red <;> assumption
+
+omit [Preorder α] [Ord α] [LawfulOrd α] in
+@[simp]
+theorem childInv_black_node {l r : Raw α} :
+    ChildInv (.node l d .black r) ↔ ChildInv l ∧ ChildInv r := by
+  constructor
+  · intro h
+    cases h
+    simp_all
+  · rintro ⟨_, _⟩
+    apply ChildInv.black <;> assumption
+
+omit [Preorder α] [Ord α] [LawfulOrd α] in
+@[aesop safe apply]
+theorem childInv_node {l r : Raw α} (h : ChildInv (.node l d c r)) :
+    ChildInv l ∧ ChildInv r := by
```

Comment:
Somewhat weird apply rule as you might accidentally construct a tree that's the wrong way around. Though you could register this as a forward rule I guess.